### ISSUE TITLE
Unable to execute decompiler on Mac

### ISSUE THREAD

**Issue by HarukaMa** [avast-tl/retdec-idaplugin#29]
I tried to run the plugin on Mac. Plugin load seems okay but when I tried to ^D on a function, I only get the error:

```Loading of output C file FAILED.```

The log stops at `[RetDec info]   :	Running the decompilation command ...`, and manually executing the printed command above works and produces the expected output file.

I've tried to add some debug code, and it seems that `std::system()` call returns 0 immediately. I have tried to modify the shebang of `retdec-decompiler.sh` to make sure that it's using current bash. I've tried to manually run a code snippet to test it but it successfully executed the decompiler script:

```cpp
#include <cstdlib>
#include <string>
int main() {

std::string a = std::string("'/usr/local/bin/retdec-decompiler.sh' ..."); // omitted parameters here
std::system(a.c_str());
return 0;
}
```

**Comment by Maroc-OS** [comment]
Because bash script doesn't work great on macOS.

Try python branch.

BTW: what is the output of : 

`/usr/local/bin/retdec-decompiler.sh --help`

**Comment by HarukaMa** [comment]
```
$ /usr/local/bin/retdec-decompiler.sh --help
Decompiles the given file into the selected target high-level language.

Usage:
    /usr/local/bin/retdec-decompiler.sh [ options ] file

Options:
...(lists all options available)
```

It seems `bash-to-python` branch is actually behind `master` though.

**Comment by PeterMatula** [comment]
Looks like another problem with Bash. We can try to get back to this after Bash is replaced with Python implementation (soon I hope) - hopefully, it will be fixed.

**Comment by PeterMatula** [comment]
Can you try it now after we got rid of Bash?

**Comment by HarukaMa** [comment]
It worked, but only after changing `/usr/bin/python` to point that to python 3. It seems when we are in `RdGlobalInfo::isDecompilerInSpecifiedPath` the `pythonCmd` is `python`, `initPythonCommand` was not being run to set it to `python3`. I think `/usr/bin` normally could not be modified because of SIP, so the plugin would stuck to python 2 which will fail to run the decompiler.

**Comment by PeterMatula** [comment]
`pythonCmd` has no default value, `initPythonCommand()` should always run in plugin initialization and set it up. If it does not, it is a bug. But I think it is very likely it did run and set up `pythonCmd` to `python`.
* How did you figure out it did not run?
* Does `python3 --version` or `py -3 --version`  work in your command line? If not, `python` is used.

**Comment by HarukaMa** [comment]
I added some info message output in corresponding functions, and they are not printed out. Will try to confirm that later.

python3 works on my command line.

**Comment by HarukaMa** [comment]
Sorry if the email of this reply have different content, I missed one output. Please read the new version below.

I made some very crude logs to check if they were run:

```diff
@@ -116,8 +116,10 @@ bool RdGlobalInfo::isSelectiveDecompilation()
  */
 bool RdGlobalInfo::initPythonCommand()
 {
+    INFO_MSG("Z");
        if (runCommand("python3", "--version") == 0)
        {
+           INFO_MSG("A");
                pythonCmd = "python3";
                return false;
        }
@@ -137,6 +139,7 @@ bool RdGlobalInfo::initPythonCommand()

 bool RdGlobalInfo::isDecompilerInSpecifiedPath() const
 {
+    INFO_MSG("%s\"%s\" --help\n", pythonCmd.c_str(), decompilerPyPath.c_str());
        return runCommand(pythonCmd, "\"" + decompilerPyPath + "\" --help") == 0;
 }
```

Only `Z` is present in the log, so it didn't find the python3 installation (maybe because of the binary location?)

```
# MrX @ harukanomacbook-pro in ~/src/retdec-idaplugin on git:master x [16:20:33]
$ python3 --version
Python 3.6.2

# MrX @ harukanomacbook-pro in ~/src/retdec-idaplugin on git:master x [16:20:35]
$ which python3
/usr/local/bin/python3
```

**Comment by PeterMatula** [comment]
Ok, so `initPythonCommand()` is run, but it does not successfully execute `python3 --version` even though you can execute it from the terminal.

Is it possible that the PATH is somehow modified in order to find `python3` only in your terminal but not globally - so the mechanism in plugin does not see it?

First, can you try to modify `initPythonCommand()` like this?
```cpp
if (runCommand("/usr/local/bin/python3", "--version") == 0)
{
	INFO_MSG("A");
	pythonCmd = "/usr/local/bin/python3";
	return false;
}
```
If this works then `python3` can be executed and used, but plugin environment simply does not see it. If it does not, then we have some bigger problem.

Second (if first works), `runCommand()` is using IDA SDK's `launch_process()` mechanism, and I don't really know how it works internally. Could you try to replace it with `std::system()` in `initPythonCommand()`. First
```cpp
if (std::system("/usr/local/bin/python3 --version") == 0)
{
	INFO_MSG("A");
	pythonCmd = "/usr/local/bin/python3";
	return false;
}
```
and then:
```cpp
if (std::system("python3 --version") == 0)
{
	INFO_MSG("A");
	pythonCmd = "python3";
	return false;
}
```

If both work (mainly `"python3 --version"`), then there is some difference between `std::system()` and `launch_process()`. If only `"/usr/local/bin/python3"` works, then whatever environment is plugin using, the PATH is not set in a way `python3` can be found.

**Comment by HarukaMa** [comment]
After testing I've found that it is caused by global PATH. For applications launched outside of terminals, the PATH needs to be changed using `launchctl setenv PATH ...` to make applications see them. After setting the PATH in `launchctl` the plugin could find `python3` successfully.

**Comment by PeterMatula** [comment]
So, in 05c7c47 i did the following:
* Python interpreter-related checks are not done at plugin initialization. For unknown reason it was crashing IDA on Windows (on Linux, it was ok).
* Python interpreter-related checks are done before each decompilation.
   * Cons: More checks, more external command invocations from plugin. Should not be a problem.
   * Pros: No crashes - safer (I never know what can and cannot be done in plugin's init function). Users are not bothered at every IDA start if they have a misconfigured plugin but don't use it. Checks can react to the current system state - state can change during one IDA session.
* New check for python interpreter version - must be at least 3.4, otherwise decompilation will not be attempted. Therefore if `python` is used, and it is Python 2, as it was in your case, you would be notified. Also `PATH` is printed in the warning message, so it would be clearer what happened.
* It is possible to configure path to the Python interpreter in plugin's config window -  you do not have to hack your system just for RetDec IDA plugin. Once configured, it will be remembered.

I'm closing this as solved. You can try it and if there is a problem re-open, or open a new issue. I'm still solving some problems with empty warning windows on Windows, it is caused by some code related to this, but the issue is not really related to the original problem, so this can be closed.

**Comment by securebluefox** [comment]
I want to know how to solve this problem, I still get this error.
![](http://img.lhyerror404.cn/error404/2020-03-02-122341.png)
![](http://img.lhyerror404.cn/error404/2020-03-02-122445.png)
My system is Mac os 10.15

**Comment by s3rvac** [comment]
@hebtuerror404 Please, create a [new issue](https://github.com/avast/retdec-idaplugin/issues/new) as this one is closed and very old. Also, attach please the input file and the out from IDA's console.
